WordPress/gutenberg
# Query pagination with InnerBlocks

## Description
Resolves: https://github.com/WordPress/gutenberg/issues/26557 - other follow ups for improvements will come.

In this PR `Query Pagination` block is implemented with InnerBlocks and can support multi query pagination in the same page. The only allowed blocks inside it are three new blocks:
1. `QueryPaginationNext`- Displays a next posts link
2. `QueryPaginationPrevious` - Displays a previous posts link
3. `QueryPaginationNumbers` - Displays numbered pagination

In `QueryPaginationNext` and `QueryPaginationPrevious` you can edit the labels and/or background/link color. I haven't added many options yet, as this PR is big as is and beautiful things can be achieved for now with just a couple of lines of css, like adding some padding to these blocks or changing the `justify-content` css property in `QueryPagination` block.



## Testing Instructions
1. With a FSE theme go to (or create) an `archive` template and add Query blocks
2. You could also add more `custom Query` blocks and in each of them you should add a `QueryPagination` block (or more if you like, ex on top and bottom)
3. If you want to test custom Queries, don't forget to toggle the `Inherit query from URL` option, that is by default `true`
4. Then visit an `archive` page (ex a `category's` page) and click pagination links.
5. Observe that multi query pagination is correct and that it respects the `max pages` param of the `Query` block

## Alternative Testing in pages - doesn't include the Queries that inherit from global context.
1. Insert `Query` and `QueryPagination` blocks (don't forget to toggle the `Inherit query from URL` option)
2. Test multi query pagination with custom queries only. 

## Notes
1. See breaking change below.
2. I have created a new file for `Query utils` in `lib` and I'm not sure if this is how the file should be named or in that place. The problem is that I needed a function that constructs a WP_Query args object from a `Query` block properties. This is needed and used by `QueryLoop`, `QueryPaginationNumbers` and `QueryPaginationNext` blocks in order to work properly.
3. There is no `live` preview in the editor, as there are context detection problems yet and to be honest I'm not sure it solves more problems if there was. For example you I think it's more important to place/style the `links` (like the Previous posts) even if you are previewing the first page of your query..

## Screenshots 
<img width="411" alt="Screenshot 2021-01-12 at 10 25 34 AM" src="https://user-images.githubusercontent.com/16275880/104288754-e3f6fa00-54c0-11eb-9599-c870d2f067ac.png">

![queryPagination](https://user-images.githubusercontent.com/16275880/104289884-53b9b480-54c2-11eb-8068-fe202e9e9682.gif)



## Types of changes
This PR breaks previous `QueryPagination`, but I haven't created a deprecation as it wasn't working properly before and I don't think it was used, besides the experimental nature of it.

## Checklist:
